### DISCUSSION TITLE
Is there a better way? (api abstraction or filler)

### DISCUSSION THREAD

**Discussion by albertoVieiraNeto** [seaweedfs/seaweedfs#4216] [Q&A]
I intend to use seaweedfs as the backend storage of a few internal apis, the thing is i have a few necessities

- it must authenticate against oidc.
- it must control user access to resources.
- it must have a api/route that would allow me to list, order, filter and page all the contents of a bucket while paging, in a flatened manner.

by what i read, seaweadfs does not offer any sort of access control or integration with a third party authentication, it also does not offer the listing api for several reasons already discussed in several issues.

my question is, im thinking of building a api on top of seaweed that would intercept all of the s3 gateway calls (only intend to use the s3) cache them in a diferent database  tables (postgres?) with a few diferent tables optimized for filtering and also add a oidc based access control.

i also intent to abstract away the fid, using this cache and only offer a Guid back to the consumers.

is there a better way, using the resources that im not seeing?

Note:
looking at the architecture, if a were able to get the "user" (header) in a filler, i think would be able to implement almost all of this there, but looking at it im not sure of the lifecycle if a were to have multiple servers for HA.

**Comment by chrislusf** [comment]
The direction is good. Access control can get complicated. Not sure about the exact details.
